linq4j/src/main/java/org/apache/calcite/linq4j/tree/OptimizeShuttle.java

```diff
@@ -164,6 +169,9 @@ && eq(cmp.expression1, expression2)) {
       Expression expression1) {
     //
     Expression result;
+    if (expression0 instanceof ParameterExpression) {
```

Comment:
Could you elaborate on why we need this optimization here? I just removed them and find `testOptimizeBinaryNullCasting` can still pass

Reply:
this is not a "nitpicking" this is my fail, append additional test. thanks !